internal/functions/efgs/publish-to-key-server.go

```diff
@@ -0,0 +1,337 @@
+package efgs
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	efgsutils "github.com/covid19cz/erouska-backend/internal/functions/efgs/utils"
+	"github.com/covid19cz/erouska-backend/internal/logging"
+	"github.com/covid19cz/erouska-backend/internal/utils"
+	keyserverapi "github.com/google/exposure-notifications-server/pkg/api/v1"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+var seededRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+var client = http.Client{}
+
+//BatchSize Max. count of exposure keys in a single batch.
+const BatchSize = 30
+
+//PublishKeysToKeyServer Publish exposure keys to Keys server.
+func PublishKeysToKeyServer(ctx context.Context, haid string, keys []keyserverapi.ExposureKey) error {
+	logger := logging.FromContext(ctx)
+
+	keyServerConfig, err := utils.LoadKeyServerConfig(ctx)
+	if err != nil {
+		logger.Fatalf("Could not load key server config: %v", err)
+		return err
+	}
+
+	verificationServerConfig, err := utils.LoadVerificationServerConfig(ctx)
+	if err != nil {
+		logger.Fatalf("Could not load verification server config: %v", err)
+		return err
+	}
+
+	if len(keys) > BatchSize {
+		batches := splitKeys(keys, BatchSize)
+		for _, batch := range batches {
+			resp, err := signAndPublishKeys(ctx, verificationServerConfig, keyServerConfig, haid, batch)
+			if err != nil {
+				logger.Errorf("Error when publishing keys: %v", err)
+				return err
+			}
+
+			logger.Infof("Batch of %v keys uploaded (%v sent), going on", resp.InsertedExposures, len(batch))
+		}
+	} else {
+		// single batch
+		_, err = signAndPublishKeys(ctx, verificationServerConfig, keyServerConfig, haid, keys)
+		if err != nil {
+			logger.Errorf("Error when publishing keys: %v", err)
+			return err
+		}
+	}
+
+	logger.Info("Keys uploaded to Key server")
+
+	return nil
+}
+
+func signAndPublishKeys(ctx context.Context, verificationServerConfig *utils.VerificationServerConfig, keyServerConfig *utils.KeyServerConfig, haid string, keys []keyserverapi.ExposureKey) (*keyserverapi.PublishResponse, error) {
+	logger := logging.FromContext(ctx)
+
+	vc, err := requestNewVC(ctx, *verificationServerConfig)
+	if err != nil {
+		logger.Errorf("Error when getting VC: %v", err)
+		return nil, err
+	}
+
+	token, err := verifyCode(ctx, *verificationServerConfig, vc)
+	if err != nil {
+		logger.Errorf("Error when getting token: %v", err)
+		return nil, err
+	}
+
+	hmacKey := make([]byte, 16)
+	_, _ = seededRand.Read(hmacKey)
+
+	certificate, err := getCertificate(ctx, *verificationServerConfig, keys, token, hmacKey)
+	if err != nil {
+		logger.Debugf("Error when getting certificate: %v", err)
```

Comment:
`logger.Errorf(...` shoud be better. This error is on the same level as error on line 77.

Reply:
This is intentional - here it's logged as debug but it's propagated to the upper level, where it's logged again as failure as error. So, actually, lines 71 and 77 are IMHO wrong :-)  
Fixed now.